### ISSUE TITLE
alacritty: implement built-in box-drawing characters

### ISSUE THREAD

**Issue by marceline-cramer** [hearth-rs/hearth#39]
[Box-drawing characters](https://en.wikipedia.org/wiki/Box-drawing_character)
are special Unicode characters used for drawing shapes in terminal interfaces.
Most fonts don't come with glyphs for these characters, so it's typically up to
the terminal emulator to draw these characters instead.

Without implementing this, many terminal programs will fail to display properly
because of missing glyphs. This is moderately high-priority to correct with
regards to Hearth's development.

To implement this, we can reference the built-in glyphs generated by
[Alacritty](https://github.com/alacritty/alacritty/blob/master/alacritty/src/renderer/text/builtin_font.rs).
A major difference between Alacritty's glyph rendering and Hearth's is that it
renders its glyphs to GPU-ready textures but Hearth uses meshes for non-MSDF
glyphs. This will need to be kept in mind while designing our procedural
glyph generation code. We will also need to introduce code to take glyphs for
specific characters from the built-in code but glyphs from the loaded font
otherwise.

Although this is moderate priority and high difficulty, implementing this will
not require much graphics-specific knowledge. Everything can be done CPU-side
with no need to modify any GPU-related code.

**Comment by marceline-cramer** [comment]
A friend suggested that Nerd Font Symbols may be useful for generalized glyph fallbacks.

**Comment by marceline-cramer** [comment]
[Noto Sans Mono](https://notofonts.github.io/noto-docs/specimen/NotoSansMono/) fits our criteria as a good fallback font that includes box drawing and block element Unicode glyphs.

`/usr/share/fonts/noto/NotoSansMono-Regular.ttf` is only ~600KB on my system, so it wouldn't add a lot of bloat to binary distributions.

**Comment by marceline-cramer** [comment]
Ok, so I ran the `rend3-alacritty` demo with Noto Sans Mono and while there are some box-drawing and block element characters that are working as intended, not all of them are.

Here's `notcurses-demo jungle` in rend3-alacritty:
![20230823_00h33m22s_grim](https://github.com/hearth-rs/hearth/assets/56894066/5bc5cd61-df17-4cbd-8c8d-ff9879dea792)

For comparison, here's `notcurses-demo jungle` in foot:
![20230823_00h33m29s_grim](https://github.com/hearth-rs/hearth/assets/56894066/21db3ff3-051c-4bad-a326-91a1059e2986)

So what's missing?

**Comment by marceline-cramer** [comment]
Because the missing characters are colored magenta, I hypothesize that the glyphs are fine and what's happening here is that `rend3-alacritty` is defaulting to magenta for indexed colors.

We can use Alacritty for reference on how to generate indexed colors here: https://github.com/alacritty/alacritty/blob/master/alacritty/src/display/color.rs

**Comment by marceline-cramer** [comment]
So I attempted to add indexed color support (and it was very good that I did), but the problem was that I was not using the custom colors that the terminal emulator overwrote my generated colors with. The jungle demo was using a palette that overwrote those indexed colors. Factoring in the custom colors fixed the issue:

![20230823_12h45m05s_grim](https://github.com/hearth-rs/hearth/assets/56894066/c2163d55-a9df-4848-bbd5-06544208eac2)

**Comment by marceline-cramer** [comment]
Ok I performed a bunch of fixes including better color management in #156.

Seems like mononoki (which is what we're already including in the source tree) already defines box-drawing and block shapes. However, Braille glyphs are not present!

What do we do next?

**Comment by siltyy** [comment]
There's always the option of adding an extra build step that strips all glyphs present in mononoki out of Noto Sans Mono and either combines those two fonts or using one as a fallback for the other. Avoids any costly overlap, while still ensuring that *most* stuff will render properly.

**Comment by marceline-cramer** [comment]
Hmmm, I for one definitely prefer mononoki out over Noto Sans Mono. Is there any other reason to do that other than to use Noto Sans Mono glyphs?

**Comment by lilithium-hydride** [comment]
Unless I'm misunderstanding, the issue is that mononoki doesn't *have* all the desired glyphs. Unless we can somehow get the font designer to support those glyphs, then we don't have any choice but to in some way fall back on another font for them. It doesn't have to be Noto Sans Mono, I only mentioned that one specifically because you had brought it up earlier.

**Comment by siltyy** [comment]
Unless I'm misunderstanding, the issue is that mononoki doesn't have all the desired glyphs. Unless we can somehow get the font designer to support those glyphs, then we don't have any choice but to in some way fall back on another font for them. It doesn't have to be Noto Sans Mono, I only mentioned that one specifically because you had brought it up earlier.

**Comment by marceline-cramer** [comment]
I don't think that Noto Sans Mono has Braille glyphs either. :( Both mononoki and NSM support box-drawing and block characters, but NOT Braille glyphs.

If you could find some options for fonts that DO support Braille glyphs, it would be a big help.

**Comment by siltyy** [comment]
[It seems to](https://github.com/search?q=repo%3Anotofonts%2Fsymbols+path%3A%2F%5Esources%5C%2FNotoSansSymbols2%5C.glyphspackage%5C%2Fglyphs%5C%2F%2F+braille&type=code), but interestingly enough, I can't seem to find the glyphs on my system's own copy of the font.

**Comment by siltyy** [comment]
Some fonts with braille support are also listed [here](https://www.fileformat.info/info/unicode/char/2801/fontsupport.htm). I'd use the site to check the fonts on my own system but I'm not attempting to set up Flash Player in 2023.

**Comment by marceline-cramer** [comment]
I think the best way to handle this issue now is to open a new issue to support multiple guest-selected font fallbacks for terminals so that we can delegate the option of which Braille-supporting font to load to the guest. I'll get on that whenever I have the time.
